llmfoundry/callbacks/async_eval_callback.py

```diff
@@ -234,76 +209,250 @@ def __init__(
+                interval = Time(i['value'], TimeUnit(i['unit']))
+                self.checkpoints_evaled[interval] = (c, rn)
+
+            log.info(
+                f'Loaded previous checkpoints evaled: {self.checkpoints_evaled}'
+            )
+
+    @staticmethod
+    def _get_ready_sharded_checkpoints(
+        checkpointer_checkpoints: Dict[str, Timestamp],
+        remote_files: List[str],
+    ) -> Dict[str, Timestamp]:
+        """Identify checkpoints ready to be evaled based on remote files.
+
+        This has special logic for sharded checkpoints to consider checkpoints composed
+        of multiple shards (one per gpu) and metadata
+
+        Args:
+            checkpointer_checkpoints: All checkpoints from the checkpointer state
+            remote_files: List of remote files in the save folder
+
+        Returns:
+            Dict of checkpoints that are complete and ready to be evaled
+        """
+        # Count the number of shards for each checkpoint group
+        remote_file_group_counts = Counter()
+        for f in remote_files:
+            interval_path = Path(f).parts[-2]
+            remote_file_group_counts[interval_path] += 1
+
+        # Check if all shards are present for each checkpoint group
+        checkpoints_to_eval = {}
+        for checkpoint, interval in checkpointer_checkpoints.items():
+            # eg {save_folder}/ep0-ba1/.
+            interval_path = Path(checkpoint).parts[-2]
+
+            # expecting one shard per gpu + 1 for metadata
+            expected_shard_count = dist.get_world_size() + 1
+            if remote_file_group_counts[interval_path] != expected_shard_count:
+                log.debug(
+                    f'Checkpoint {checkpoint} not fully uploaded (missing shards '
+                    +
+                    f'{remote_file_group_counts[interval_path]}/{expected_shard_count}), skipping'
+                )
+                continue
+
+            checkpoints_to_eval[interval_path] = interval
+
+        return checkpoints_to_eval
+
+    @staticmethod
+    def _get_ready_single_checkpoints(
+        checkpointer_checkpoints: Dict[str, Timestamp],
+        remote_checkpoints: List[str],
+    ) -> Dict[str, Timestamp]:
+        """Identify checkpoints ready to be evaled based on remote checkpoints.
+
+        This is much simpler than the sharded case, because there is only one file
+
+        Args:
+            checkpointer_checkpoints: All checkpoints from the checkpointer state
+            remote_checkpoints: List of remote checkpoints in the save folder
+
+        Returns:
+            Dict of checkpoints that are complete and ready to be evaled
+        """
+        unique_remote_checkpoints = set(remote_checkpoints)
+
+        checkpoints_to_eval = {}
+        for checkpoint, interval in checkpointer_checkpoints.items():
+            # eg {save_folder}/ep0-ba1-rank0.pt
+            interval_path = Path(checkpoint).parts[-1]
```

Comment:
also i think technically the time information doesn't have to be in the filename and could be just in the path...if this is hard its probably ok not to optimize for now.

Reply:
This info doesn't actually need to be in the path because we use the checkpoint ts (interval) from the checkpointer, the comment was just an example from parsing.

For this and above, it does assume `save_filename` in train.py is unique by checkpoint ts and does not include a directory (such as `foobar/ep{epoch}-ba{batch}-rank{rank}.pt`).. Does this seem reasonable? Can add an error on callback initialization to check for this not being the case

This makes the `load_path` logic for launching the eval runs significantly easier because it will be `{save_folder}/{what is referred to as checkpoint}` from remote.

Reply:
ok yeah as long as you hard error if save_filename has a folder, im good with it.